File: server/src/progression.ts
// PostgreSQL integer XP stays exact in JavaScript and Kotlin Long.
export function progression(totalXp: number) {
  const level = Math.floor((1 + Math.sqrt(1 + totalXp / 25)) / 2);
  const currentLevelStartXp = 100 * level * (level - 1);
  const nextLevelXp = 100 * level * (level + 1);
  return { totalXp, level, currentLevelStartXp, nextLevelXp,
    xpIntoCurrentLevel: totalXp - currentLevelStartXp,
    xpRequiredForNextLevel: 200 * level };
}

export function progressionResult(previousTotalXp: number, newTotalXp: number, xpAwarded: number) {
  const previous = progression(previousTotalXp);
  const next = progression(newTotalXp);
  return { ...next, xpAwarded, previousTotalXp, newTotalXp,
    previousLevel: previous.level, newLevel: next.level,
    previousXpIntoCurrentLevel: previous.xpIntoCurrentLevel,
    previousXpRequiredForNextLevel: previous.xpRequiredForNextLevel };
}

File: android/app/src/main/java/com/droidnova/mathfight/profile/Progression.kt
package com.droidnova.mathfight.profile

import org.json.JSONObject

// All progression values come from the server; Android only renders them.
data class PlayerProgression(
    val totalXp: Long, val level: Int, val currentLevelStartXp: Long, val nextLevelXp: Long,
    val xpIntoCurrentLevel: Long, val xpRequiredForNextLevel: Long
) {
    val fraction: Float get() = (xpIntoCurrentLevel.toDouble() / xpRequiredForNextLevel).toFloat().coerceIn(0f, 1f)
}

data class XpResult(
    val matchId: String, val xpAwarded: Long, val previousTotalXp: Long, val newTotalXp: Long,
    val previousLevel: Int, val newLevel: Int, val previousFraction: Float,
    val progression: PlayerProgression
)

fun parseProgression(value: JSONObject): PlayerProgression? {
    val total = value.optLong("totalXp", -1)
    val level = value.optInt("level", 0)
    val start = value.optLong("currentLevelStartXp", -1)
    val next = value.optLong("nextLevelXp", -1)
    val into = value.optLong("xpIntoCurrentLevel", -1)
    val required = value.optLong("xpRequiredForNextLevel", -1)
    if (total < 0 || level < 1 || start < 0 || next <= start || into < 0 || required <= 0) return null
    return PlayerProgression(total, level, start, next, into, required)
}

fun parseXpResult(matchId: String, value: JSONObject): XpResult? {
    val progression = parseProgression(value) ?: return null
    val previous = value.optLong("previousTotalXp", -1)
    val awarded = value.optLong("xpAwarded", -1)
    val newTotal = value.optLong("newTotalXp", -1)
    val previousLevel = value.optInt("previousLevel", 0)
    val newLevel = value.optInt("newLevel", 0)
    val required = value.optLong("previousXpRequiredForNextLevel", 0)
    val into = value.optLong("previousXpIntoCurrentLevel", -1)
    if (previous < 0 || awarded < 0 || newTotal < previous || previousLevel < 1 || newLevel < previousLevel || required <= 0 || into < 0) return null
    return XpResult(matchId, awarded, previous, newTotal, previousLevel, newLevel,
        (into.toDouble() / required).toFloat().coerceIn(0f, 1f), progression)
}
